stdlib/asm/collections/sorted_array.masm

```diff
@@ -0,0 +1,279 @@
+        #=> [is_value_found, maybe_value_ptr, start_ptr, end_ptr]
+    else
+        # the value was not found
+        
+        # pre-compute is_start_ptr, is_end_ptr to reduce the depth of the generated MAST
+        # (6 cycles)
+        dup.6 dup.1 eq
+        #=> [is_end_ptr, maybe_value_ptr, VALUE, start_ptr, end_ptr]
+
+        dup.6 dup.2 eq
+        #=> [is_start_ptr, is_end_ptr, maybe_value_ptr, VALUE, start_ptr, end_ptr]
+
+        if.true
+            # the value was not found, and maybe_value_ptr = start_ptr
+
+            if.true
+                # the list is empty
+                #=> [maybe_value_ptr = start_ptr = end_ptr, VALUE, start_ptr, end_ptr]
+
+                # Prepare output (6 cycles)
+                movdn.4 dropw push.0
+                #=> [is_value_found = 0, value_ptr = start_ptr, start_ptr, end_ptr]
+            else
+                # the value was not found, maybe_value_ptr = start_ptr, and the list is not empty
+                # assert that the value at start_ptr > VALUE
+                #=> [maybe_value_ptr = start_ptr, VALUE, start_ptr, end_ptr]
+
+                movdn.4
+                #=> [VALUE, maybe_value_ptr, start_ptr, end_ptr]
+
+                # dereference maybe_value_ptr (7 cycles)
+                dup.4 padw movup.4 mem_loadw
+                #=> [MAYBE_VALUE, VALUE, maybe_value_ptr, start_ptr, end_ptr]
+
+                # there was no match, the first array element must be larger than VALUE
+                # (123 cycles)
+                exec.word::lt
+                assert.err="lowerbound_array invariant: start_ptr not greater than VALUE"
+
+                # No value was found (1 cycle)
+                push.0
+                #=> [is_value_found = 0, maybe_value_ptr = start_ptr, start_ptr, end_ptr]
+            end
+        else
+            if.true
+                # the value was not found, maybe_value_ptr = end_ptr
+                # the list is not empty, otherwise maybe_value_ptr = start_ptr and handled above
+                # assert that the value before end_ptr < VALUE
+
+                #=> [maybe_value_ptr = end_ptr, VALUE, start_ptr, end_ptr]
+
+                # fetch the last word in the array (10 cycles)
+                dup movdn.5 sub.4 padw movup.4 mem_loadw
+                #=> [LAST_WORD, VALUE, maybe_value_ptr, start_ptr, end_ptr]
+
+                # there was no match, the last array element must be smaller than VALUE
+                # (119 cycles)
+                exec.word::gt
+                assert.err="lowerbound_array invariant: last word not smaller than VALUE"
+
+                # No value was found (1 cycle)
+                push.0
+                #=> [is_value_found = 0, maybe_value_ptr = end_ptr, start_ptr, end_ptr]
+            else
+                # The value was not found, and the maybe_value_ptr is neither the start nor the end
+                # of the list. Make sure that the value at maybe_value_ptr-1 < VALUE and at maybe_value_ptr > VALUE
```

Comment:
nit: same comment here and would probably write them as two bullet points i.e., 

... make sure that 
1. the value at maybe_value_ptr-1 is strictly less than VALUE, and that
2. the value at maybe_value_ptr is strictly greater than VALUE

Reply:
Agreed about bullet points. See my comment about logical relationships: https://github.com/0xMiden/miden-vm/pull/2114#discussion_r2348341757